copy/copy.go

```diff
@@ -655,17 +696,108 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, po
+		// return if the image destination does not blob locking.
+		tryReusingBlob := func() (bool, types.BlobInfo, digest.Digest, error) {
+			reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+			if err != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+			}
+			if reused {
+				logrus.Debugf("Skipping blob %s (already present):", srcInfo.Digest)
+				bar.SetTotal(0, true)
+				ic.c.createProgressBar(pool, srcInfo, "blob", "skipped: already exists", mpb.BarReplaceOnComplete(bar))
+				return true, blobInfo, cachedDiffID, nil
+			}
+
+			if !ic.c.dest.SupportsBlobLocks() {
+				return false, types.BlobInfo{}, "", nil
+			}
+
+			// Check if another process might be copying the layer at the
+			// moment.  If so, wait for it's completion and try to reuse
+			// the blob again.
+			//
+			logrus.Debugf("Checking if blob %s is already being copied by another process", srcInfo.Digest)
+
+			// LockBlob() does not support try-lock semantics which would
+			// be a straight-forward mechanism to detect if another process
+			// is currently copying a specific layer.  However, with
+			// LockBlob() we are constrained by the file locks in
+			// containers/storage which cannot support try-lock semantics
+			// as fcntl(2) does not support that.
+			//
+			// The trick we are using here is to acquire the lock in a
+			// separate goroutine, which sets a boolean signal-variable as
+			// soon as the lock is acquired.  The main-routine is then
+			// sleeping for a certain amount of time.  If the lock has not
+			// been acquired until then, the progress bar will be updated
+			// with a message indicating that the blob is being copied by
+			// another process and wait until the signal-variable is set
+			// true.
+			acquiredBlobLock := false
+			go func() {
+				ic.c.dest.LockBlob(srcInfo)
+				acquiredBlobLock = true
+			}()
+
+			if !acquiredBlobLock {
+				// The initial sleep time is `progressBarRefreshRate`
+				// to make sure the bar is properly rendered before
+				// replacing it.
+				time.Sleep(progressBarRefreshRate)
+				if !acquiredBlobLock {
+					logrus.Debugf("Blob %s is already being copied by another process", srcInfo.Digest)
+					// If we still don't have the lock, replace the
+					// initial progress bar with a new one
+					// indicating that the blob is copied by
+					// another process.
+					//
+					// Note that we call SetTotal() after the new
+					// bar is created to avoid potential flickering
+					// if the bar removal and replacement happens
+					// across refresh rates.
+					toReplaceBar := bar
+					bar = ic.c.createProgressBar(pool, srcInfo, "blob", "stopped: copied by another process", mpb.BarRemoveOnComplete(), mpb.BarReplaceOnComplete(bar))
+					toReplaceBar.SetTotal(0, true)
+
+					for !acquiredBlobLock {
+						time.Sleep(50 * time.Millisecond)
```

Comment:
Can’t this use a channel to block instead?

That would also allow:
```go
lockingDoneChan := startALockingGoroutine()
weAreProbablyBlockedChan := startATimerForRefreshInterval
select {
case <- lockingDoneChan: // We are the only one, continue as usual:
case <- weAreProbablyBlockedChan: // We are blocked
    // tell the user
    <- lockingDoneChan // Block on the lock
}
// either way, lockingDoneChan has been waited on here
// (FIXME: do we need to clean up weAreProbablyBlockedChan and the timer? channels can be tricky.)
```

which does not require waiting for the whole refresh interval if we can acquire the lock faster.

Reply:
Sure, we can do that. Usually, I am hesitant using channels as they seem to be a common source for bugs but this case seems small enough to brain-parse the code.

Reply:
This is still outstanding.

Reply:
We're now doing this via semaphore which has a `TryAcquire`.